src/language/from_relay/mod.rs

```diff
@@ -1472,6 +1566,367 @@ fn compile_expression(
+                    {
+                        let upsampling_id = glenside_expr.add(Language::RelayOperator(
+                            crate::language::RelayOperator::RelayUpSampling,
+                        ));
+                        glenside_expr.add(Language::RelayOperatorCall(
+                            vec![upsampling_id, data_id, scale_h_id, scale_w_id, layout_id]
+                                .into_boxed_slice(),
+                        ))
+                    } else {
+                        todo!()
+                    }
+                }
+                "concatenate" => {
+                    assert_eq!(call.args.len(), 1);
+                    let attrs = call
+                        .attrs
+                        .clone()
+                        .downcast::<tvm::ir::relay::attrs::transform::ConcatenateAttrs>()
+                        .unwrap();
+
+                    assert!(attrs.axis >= 0);
+
+                    let tuple = call
+                        .args
+                        .get(0)
+                        .unwrap()
+                        .clone()
+                        .downcast::<tvm::ir::relay::Tuple>()
+                        .ok()
+                        .unwrap();
+                    let mut concatted_id = get_compiled_expression(tuple.fields.get(0).unwrap());
+                    for i in 1..tuple.fields.len() {
+                        // TODO(@gussmith23) Layout assumption
+                        let to_be_concatted_id = get_compiled_expression(
+                            tuple.fields.get(i.try_into().ok().unwrap()).unwrap(),
+                        );
+                        concatted_id = access_concatenate(
+                            glenside_expr,
+                            concatted_id,
+                            to_be_concatted_id,
+                            attrs.axis as usize,
+                        );
+                    }
+
+                    concatted_id
+                }
+                "reshape" => {
+                    assert_eq!(call.args.len(), 1);
+                    let attrs = call
+                        .attrs
+                        .clone()
+                        .downcast::<tvm::ir::relay::attrs::transform::ReshapeAttrs>()
+                        .unwrap();
+                    let data_id = get_compiled_expression(call.args.get(0).unwrap());
+
+                    let mut new_shape = Vec::new();
+                    for i in 0..attrs.newshape.len() {
+                        let val = attrs
+                            .newshape
+                            .get(i.try_into().ok().unwrap())
+                            .unwrap()
+                            .downcast::<IntImm>()
+                            .unwrap()
+                            .value;
+                        new_shape.push(if val < 0 {
+                            // assume -1 = 1 in reshape.newshape
```

Comment:
Is this a safe assumption? Are there checks we can add here so that we panic when the assumption is not true?

Reply:
For the models we use this assumption is true.

I think the tricky engineering bit would be to figure out how to infer the shapes since we need the shape information of its first argument. AFAIK shape information is only available after the expr is added to the egraph.

Reply:
Like we discussed, the shape information *should* be available via `checked_type`. Let me know if that's not the case!

Reply:
ah sorry the new commits aren't reflected. I decided to just use the shape information via `checked_type` instead of computing it ourselves. lmk if this is okay!